protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+deployments also prevents multiple L2s from running.
+
+The main goal is to unbundle these steps, and make them customizable and composable for better test-coverage,
+including Interop parameters specifically.
+
+## OPStackManager
+
+Specs: https://specs.optimism.io/experimental/op-stack-manager.html
+
+Older implementatipn draft: https://github.com/ethereum-optimism/optimism/pull/9985
+
+The OP-Stack manager focuses on making the deployment of the L1 contracts of an L2 chain deterministic,
+by deploying proxies to deterministic addresses (salt = chainID),
+and coupling to superchain-provided implementation contracts.
+
+This implies a different L1-deployment flow than we have today,
+and limits it to just the L2-instance specific contracts.
+A deployment phase we also want to isolate the configuration and execution of.
+
+Note that the draft sources configuration inputs explicitly, rather than coupling to the (arguably bloated) 
+deploy config.
+
+## Deployer requirements
+
+### Op-e2e needs
+
+As an op-e2e test suite, I need to:
+- Customize the deploy-config deeply, including L1 parameters and upgrade parameters, to cover all functionality.
+- Create multiple L2s with a shared L1 (to go from isolated L2s to interoperating L2s through interop hardfork).
+- Create multiple L2s with a shared L1 and shared superchain (to test interop without emulating hardfork first).
+
+### Devnet needs
+
+As a devnet user, I need to:
+- Modify the chain template, for experimentation when building new features.
+- Support different "modes" without accidental configuration mixup,
+  e.g. I may to run different hardforks, or run with different DA types.
+
+### Production needs
+
+As a production superchain host, I need to:
+- Perform a new Superchain core deployment (global contracts and implementation-contract deployments).
+- Maintain the Superchain core deployment with upgrades.
+- Not deploy/upgrade an L2 at the same time.
+
+As a production L2 chain deployer, I may need to:
+- Build an L2 on top of a pre-existing superchain core deployment.
+- Upgrade the L2 by attaching to new superchain implementation contracts.
+- Customize just the L2 part of the deployment, while retaining compatibility with the general deployment flow.
+
+Production needs may change; there are many external users and new features. Flexibility is key.
+Making the deployment more modular fits this.
+
+# Alternatives Considered
+
+## Few enshrined deploy-configs
+
+Not all that different from status-quo.
+With copy-paste we can add 1 more chain, and suffix it with `-interop`, to produce a 2-chain test environment.
+For early testing this was drafted in [PR 10777](https://github.com/ethereum-optimism/optimism/pull/10777).
+
+This does not yet unify the common superchain part of the deployment, or any of the artifacts,
+nor supports test-specific changes to parameters.
+
+## Presets and configs
+
+In Eth2 there exist two classes of configuration:
+- "preset" (named, enshrined, compile-time)
```

Comment:
I would like to make more config optional like in https://github.com/ethereum-optimism/optimism/pull/10628 to reduce footguns and make it more simple to deploy a standard config chain

Reply:
Agreed, I think we should do so through composition, to encapsulate things that are rarely modified. But not entirely freeze it either.